### PR TITLE
[opichuryn] iterators_task_2

### PR DESCRIPTION
**Pull request by opichuryn**

### CODE REVIEW

File: iterators_task_2/student/include/impl.h [hunk 1]
```diff
@@ -1,21 +1,73 @@
...
-    /** @todo rvalue writing operator*/
+    using value_type = typename Container::iterator::value_type;
+    using difference_type = typename Container::iterator::difference_type;
+    using pointer = typename Container::iterator::pointer;
+    using reference = typename Container::iterator::reference;
+    using iterator_category = std::output_iterator_tag;
 
-    /** @todo Iterator operations */
+    accumulator(Container& container)
+        : m_ref(std::ref(container))
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:15]: you can skip the std::ref here

File: iterators_task_2/student/include/impl.h [hunk 2]
```diff
@@ -1,21 +1,73 @@
...
+    }
+
+    accumulator<Container>& operator*()
+    {
+        return *this;
+    }
+
+private:
+    std::reference_wrapper<Container> m_ref;
+    typename Container::iterator m_iter;
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:66]: why do you need this iterator?

File: iterators_task_2/student/include/impl.h [hunk 3]
```diff
@@ -1,21 +1,73 @@
...
+    }
+
+    accumulator operator++(int)
+    {
+        auto tmp {*this};
+        operator++();
+        return tmp;
+    }
+
+    bool operator==(const accumulator<Container>& rhs) const
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:49]: Output iterator does not provide this operation

File: iterators_task_2/student/include/impl.h [hunk 4]
```diff
@@ -1,21 +1,73 @@
...
+    {
+        return m_ref == rhs.m_ref;
+    }
+
+    bool operator!=(const accumulator<Container>& rhs) const
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:54]: Not valid operation

File: iterators_task_2/student/include/impl.h [hunk 5]
```diff
@@ -1,21 +1,73 @@
...
+#include <functional>
 
 template <class Container>
 struct accumulator
 {
-    /** @todo Iterator traits*/
-    /** @todo Constructor with container*/
-    /** @todo lvalue writing operator*/
-    /** @todo rvalue writing operator*/
+    using value_type = typename Container::iterator::value_type;
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:8]: Does it have sense for output iterators?

File: iterators_task_2/student/include/impl.h [hunk 6]
```diff
@@ -1,21 +1,73 @@
...
+        : m_ref(std::ref(container))
+        , m_iter(container.end())
+    {}
+
+    accumulator(const accumulator<Container>& other) = default;
+
+    template <class AnotherContainer>
+    accumulator<Container>& operator=(AnotherContainer&& container)
+    {
+        std::copy(std::make_move_iterator(container.begin()),
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:24]: use std::move algorithm for this case

File: iterators_task_2/student/include/impl.h [hunk 7]
```diff
@@ -1,21 +1,73 @@
...
+    using reference = typename Container::iterator::reference;
+    using iterator_category = std::output_iterator_tag;
 
-    /** @todo Iterator operations */
+    accumulator(Container& container)
+        : m_ref(std::ref(container))
+        , m_iter(container.end())
+    {}
+
+    accumulator(const accumulator<Container>& other) = default;
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:19]: what about assignment operator?

File: iterators_task_2/student/include/impl.h [hunk 8]
```diff
@@ -1,21 +1,73 @@
...
+    template <class AnotherContainer>
+    accumulator<Container>& operator=(AnotherContainer&& container)
+    {
+        std::copy(std::make_move_iterator(container.begin()),
+                  std::make_move_iterator(container.end()),
+                                      std::back_inserter(m_ref.get()));
+        return *this;
+    }
+
+    accumulator<Container>& operator=(value_type&& type)
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:30]: This adaptor works with containers, and not the values

File: iterators_task_2/student/include/impl.h [hunk 9]
```diff
@@ -1,21 +1,73 @@
...
+    {
+        m_ref.get().insert(m_iter, std::forward(type));
+        return *this;
+    }
+
+    accumulator& operator++()
+    {
+        ++m_iter;
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:38]: incrementing end iterator will result an UB, you don't need this iterator at all

File: iterators_task_2/student/include/impl.h [hunk 10]
```diff
@@ -1,21 +1,73 @@
...
+        return *this;
+    }
+
+    accumulator operator++(int)
```
Comment (smay1613) [inline iterators_task_2/student/include/impl.h:42]: not supported for single pass iterators

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: iterators_task_2/student/include/impl.h (hunk 7 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h:18]: Done

File: iterators_task_2/student/include/impl.h (hunk 1 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done

File: iterators_task_2/student/include/impl.h (hunk 2 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done, removed

File: iterators_task_2/student/include/impl.h (hunk 3 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done, removed

File: iterators_task_2/student/include/impl.h (hunk 4 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done, removed

File: iterators_task_2/student/include/impl.h (hunk 5 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done

File: iterators_task_2/student/include/impl.h (hunk 6 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h]: Done

File: iterators_task_2/student/include/impl.h (hunk 7 above)
Comment (opichuryn) [reply iterators_task_2/student/include/impl.h:18]: Done, implemented

**Pull request closed without merging**